File: app/src/main/java/com/devusercode/upchat/utils/GetTimeAgo.java
package com.devusercode.upchat.utils;

import android.app.Application;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

public class GetTimeAgo extends Application {
    /*
     * THIS CLASS IS FOR GETTING LAST SEEN TIME IN CHATS
     * IT CONVERTS TIME_STAMP INTO TIME_AGO
     */

    private static final int SECOND_MILLIS = 1000;
    private static final int MINUTE_MILLIS = 60 * SECOND_MILLIS;
    private static final int HOUR_MILLIS = 60 * MINUTE_MILLIS;
    private static final int DAY_MILLIS = 24 * HOUR_MILLIS;
    private static final int WEEK_MILLIS = 7 * DAY_MILLIS;

    public static String getTimeAgo(long time) {
        if (time < 1000000000000L) {
            // if timestamp given in seconds, convert to millis
            time *= 1000;
        }

        long now = System.currentTimeMillis();

        if (time > now || time <= 0) {
            return null;
        }

        final long diff = now - time;

        if (diff < MINUTE_MILLIS) {
            return "just now";
        } else if (diff < 2 * MINUTE_MILLIS) {
            return "a minute ago";
        } else if (diff < 50 * MINUTE_MILLIS) {
            return diff / MINUTE_MILLIS + " minutes ago";
        } else if (diff < 90 * MINUTE_MILLIS) {
            return "an hour ago";
        } else if (diff < 24 * HOUR_MILLIS) {
            return diff / HOUR_MILLIS + " hours ago";
        } else if (diff < 48 * HOUR_MILLIS) {
            return "yesterday";
        } else if (diff < WEEK_MILLIS) {
            return diff / DAY_MILLIS + " days ago";
        } else if (diff < 4L * WEEK_MILLIS) {
            if ((diff / WEEK_MILLIS) == 1) {
                return "a week ago";
            } else {
                return diff / WEEK_MILLIS + " weeks ago";
            }
        } else {
            // Display date and time if more than 1 month
            SimpleDateFormat sdf = new SimpleDateFormat("dd/MM/yyyy HH:mm", Locale.getDefault());
            return sdf.format(new Date(time));
        }
    }

    public static String parse(String time) {
        return getTimeAgo(Long.parseLong(time));
    }
}

File: app/src/main/java/com/devusercode/upchat/utils/GetTimeAgo.kt
package com.devusercode.upchat.utils

import android.app.Application
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

object GetTimeAgo : Application() {
    private const val SECOND_MILLIS = 1000L
    private const val MINUTE_MILLIS = 60 * SECOND_MILLIS
    private const val HOUR_MILLIS = 60 * MINUTE_MILLIS
    private const val DAY_MILLIS = 24 * HOUR_MILLIS
    private const val WEEK_MILLIS = 7 * DAY_MILLIS
    private const val MONTH_MILLIS = 30 * DAY_MILLIS

    fun getTimeAgo(givenTime: Long): String? {
        var time = givenTime

        if (time < 1000000000000L) {
            // if timestamp given in seconds, convert to millis
            time *= 1000
        }

        val now = System.currentTimeMillis()

        if (time > now || time <= 0) {
            return null
        }

        val diff = now - time

        return when {
            diff < MINUTE_MILLIS -> "now"
            diff < HOUR_MILLIS -> "${diff / MINUTE_MILLIS}m"
            diff < DAY_MILLIS -> "${diff / HOUR_MILLIS}h"
            diff < WEEK_MILLIS -> "${diff / DAY_MILLIS}d"
            diff < MONTH_MILLIS -> "${diff / WEEK_MILLIS}w"

            else -> {
                // Display date and time if more than 1 month
                val sdf = SimpleDateFormat("dd/MM/yyyy HH:mm", Locale.getDefault())
                sdf.format(Date(time))
            }
        }
    }

    fun parse(time: String): String? {
        return getTimeAgo(time.toLong())
    }
}